### PR TITLE
Add Scope refresh for changed or new nodes

### PR DESCRIPTION
**Pull request by TomasVotruba**

### CODE REVIEW

File: tests/Issues/IssueChangeAndIfAndDateTimeAndDeMorgan/config/configured_rule.php [hunk 1]
```diff
@@ -10,5 +10,5 @@
 return static function (RectorConfig $rectorConfig): void {
     $rectorConfig->rule(SimplifyDeMorganBinaryRector::class);
     $rectorConfig->rule(ChangeAndIfToEarlyReturnRector::class);
-    $rectorConfig->rule(DateTimeToDateTimeInterfaceRector::class);
+//    $rectorConfig->rule(DateTimeToDateTimeInterfaceRector::class);
```
Comment (samsonasik) [inline tests/Issues/IssueChangeAndIfAndDateTimeAndDeMorgan/config/configured_rule.php]: this handled at https://github.com/rectorphp/rector-src/pull/2274 for Scope not avaiable after original node verified as not Node (null)

**Comment by samsonasik** [comment]
@TomasVotruba please rebase, PR https://github.com/rectorphp/rector-src/pull/2274 handle scope not exists which seems due:

- Applied multiple rules on a Node.
- Node's original node already gone.
- Node's start token pos < 0, it still can process if the Scope is not involved, as previous recorded node is different.
- if Scope is going to be used, which node no longer has `Scope`, it can't process rule that extends `AbstractScopeAwareRector`, so that’s definitelly need a second run to apply the rule.

File: rules/DeadCode/Rector/If_/RemoveDeadInstanceOfRector.php [hunk 2]
```diff
@@ -85,7 +85,7 @@ public function getNodeTypes(): array
      * @param If_ $node
      * @return Stmt[]|null|If_
      */
-    public function refactor(Node $node): null|array|If_
+    public function refactorWithScope(Node $node, Scope $scope): ?If_
```
Comment (samsonasik) [inline rules/DeadCode/Rector/If_/RemoveDeadInstanceOfRector.php]: array can be returned at line 144

File: tests/Issues/IssueChangeAndIfAndDateTimeAndDeMorgan/config/configured_rule.php (hunk 1 above)
Comment (TomasVotruba) [reply tests/Issues/IssueChangeAndIfAndDateTimeAndDeMorgan/config/configured_rule.php]: After this PR the scope must be available everywhere. This is just temporary.

File: rules/CodeQuality/Rector/FunctionLike/RemoveAlwaysTrueConditionSetInConstructorRector.php [hunk 3]
```diff
@@ -90,81 +87,50 @@ public function go()
...
+    public function refactor(Node $node): ?array
     {
-        if ($node->stmts === null) {
+        $ifStmt = $this->matchTruableIf($node);
+        if (! $ifStmt instanceof If_) {
             return null;
         }
 
-        if ($node->stmts === []) {
+        if ($ifStmt->stmts === null) {
```
Comment (samsonasik) [inline rules/CodeQuality/Rector/FunctionLike/RemoveAlwaysTrueConditionSetInConstructorRector.php]: `$ifStmt->stmts` will never be `null`, only `[]`, checking against `null` will make next return `$ifStmts->stmts` may cause error if it is empty array `[]`

Comment (TomasVotruba) [reply rules/CodeQuality/Rector/FunctionLike/RemoveAlwaysTrueConditionSetInConstructorRector.php]: Thanks, I'll update :+1:

**Comment by TomasVotruba** [comment]
@samsonasik Btw, this is a probe PR. I'm exploring boundaries of what needs to be done, before moving to Scope refresh on every changed node. Some of those changes are related with parent/next/previous node, so I'm refactoring rules to use the parent node directly.

No need to review for time being, as it will be lot of commits just trying out different approaches. I'll ping you when I'm finished for review :+1:

**Comment by samsonasik** [comment]
@TomasVotruba ok, just for note, merge commit has quote `[Merge branch 'tv-refactor-with-scope-2' of]` need to be cleaned up to avoid broken scoped build.

**Comment by samsonasik** [comment]
@TomasVotruba please rebase due to update on handle empty stmts on RemoveAlwaysTrueConditionSetInConstructorRector https://github.com/rectorphp/rector-src/pull/2281 due to break on https://github.com/rectorphp/rector-src/pull/2280#discussion_r869776444

**Comment by samsonasik** [comment]
@TomasVotruba please rebase again, PR https://github.com/rectorphp/rector-src/pull/2288 change `MakeInheritedMethodVisibilitySameAsParentRector` back to `refactor()` as `Scope` not used as handled at `ReflectionResolver` so phpstan.neon needs update to use `refactor()` for ignore message.

**Comment by TomasVotruba** [comment]
I'm rebasing continually 🙂, no need to ping here.

**Comment by TomasVotruba** [comment]
Replaced by https://github.com/rectorphp/rector-src/pull/2292

**Pull request closed without merging**